### DISCUSSION TITLE
Does it support computed/calculated column or method

### DISCUSSION THREAD

**Discussion by bingtsingw** [romeerez/orchid-orm#58] [General]
Case one:
I have a `firstname` and `lastname` column in table, I want to calculate a `fullname` property on the fly, how can I implement it

Case two:
I have a `vipEndAt` column in the table to record the users vip end date, so I may want to have a `user.getVipInfo()` method on fly, which can calculate left dates, vip status and so on.

I searched for the docs, and found that the `model` and `repository` are focused on columns in database, there seems to miss a way to extend query resluts

**Comment by romeerez** [comment]
Case one is clear to me, currently, it is not supported, but I'm adding it to a to-do list.

For the second case, could you please elaborate? Code examples are welcome

**Reply by bingtsingw** [reply]
I came from [adonis](https://docs.adonisjs.com/guides/models/introduction) world which maps table records to a real class, so I can do anything in model class, this is a real example in my app
```typescript
class User extends BaseModel {
  @column()
  public role: string[];

  public hasRole(role: string) {
    if (this.role.includes(role)) {
      return true;
    }

    return false;
  }

  public isAdmin() {
    return this.hasRole('admin');
  }
}
```

So, in my controller, I can use this code
```typescript
const user = await User.find('xxx')
user.isAdmin() // to calculate if user has admin role
user.hasRole('editor') // to calculate if user has editor role
```

**Reply by romeerez** [reply]
I agree that it looks really neat and natural:

```ts
user.isAdmin()
user.hasRole('editor')
```

But, I rejected ActiveRecord pattern intentionally, [here is a short](https://orchid-orm.netlify.app/guide/#not-an-orm-in-oop-sense) explanation why.

Apart from making code a bit slower and less type-safe (methods cannot know if `role` was actually selected or not), I worked with Ruby on Rails before the node.js and have seen a lot of models consisting of thousands of lines of business logic, I believe it's just wrong from the beginning and MVC is the worst possible architecture. I even renamed `BaseModel` to `BaseTable` and all examples have `UserTable` rather than `UserModel`, because it's not a place for business logic but an interface to a database table.

Instead, you can define functions:

```ts
export const hasRole = (user: { roles: string[] }, role: string)  =>
  user.roles.includes(role)
  
export const isAdmin = (user: { roles: string[] }) => hasRole(user, 'admin')
```

Or make an object with role-related functions:

```ts
export const roleMethods = {
  hasRole: () => ...,
  isAdmin: () => ...,
}
```

Or do it with a class. Logic can be placed anywhere and anyhow! I got those old PTSD flashbacks when thinking of how it was done with MVC when you are forced to place all the logic into giant bloated classes, having zero type safety at the same time, surrounded by implicit magic.

**Reply by bingtsingw** [reply]
Well, I'm also came from the MVC world, so I'm more familiar with model pattern, but you make me thinking about another ways to organize logic, I'll try it in my project

**Comment by bingtsingw** [comment]
Sorry, I found `orchid` can support these cases in [repository pattern](https://orchid-orm.netlify.app/guide/orm-repo.html#kinds-of-methods), I missed the `simpleMethod` section

**Reply by romeerez** [reply]
That's right it's possible to define such methods in repo. I added that `simpleMethod` just in case, so it may be a good place for a function, but not that if it's a recommended or suggested way, only as an option. I mean, you can still define functions in any way that makes sense. I'd probably define `isAdmin` on a `service` layer (in a file `user.service.ts` or `roles.service.ts`).

**Reply by bingtsingw** [reply]
Service layer is a nice way to separate logic, but I'm happy to have an option to define it in the repo, so that I can choose one better suits my app

**Comment by bingtsingw** [comment]
Very nice ORM, I can't wait to try it in my next project.

Another suggestion: maybe you can scope this project in an organization 
`@orchid/core` `@orchid/pqb` `@orchid/orm-schema-to-zod` looks clearer and more professional

**Reply by romeerez** [reply]
`@orchid` is taken on npm, and I couldn't choose anything better than leaving it as it is.

Thank you, I honestly really like all sorts of feedbacks so feel free to create issues and ask questions!

**Comment by romeerez** [comment]
Added issue for the case one: https://github.com/romeerez/orchid-orm/issues/59
